sanitise.go

```diff
@@ -0,0 +1,192 @@
+	// dash, and dot characters.
+	UnderscoreDashDotCharacters = []rune{
+		'.',
+		'-',
+		'_'}
+
+	// AlphanumericSanitiserOpts is the options to create a sanitiser which uses
+	// the alphanumeric SanitiseFn.
+	AlphanumericSanitiserOpts = SanitiseOptions{
+		NameCharacters: ValidCharacters{
+			Ranges:     AlphanumericRange,
+			Characters: UnderscoreDashCharacters,
+		},
+		KeyCharacters: ValidCharacters{
+			Ranges:     AlphanumericRange,
+			Characters: UnderscoreDashCharacters,
+		},
+		ValueCharacters: ValidCharacters{
+			Ranges:     AlphanumericRange,
+			Characters: UnderscoreDashCharacters,
+		},
+		ReplacementCharacter: DefaultReplacementCharacter,
+	}
+)
+
+// SanitiseFn returns a sanitised version of the input string.
+type SanitiseFn func(string) string
+
+// SanitiseRange is a range of characters (inclusive on both ends).
+type SanitiseRange [2]rune
+
+// ValidCharacters is a collection of valid characters.
+type ValidCharacters struct {
+	Ranges     []SanitiseRange
+	Characters []rune
+}
+
+// SanitiseOptions are the set of configurable options for sanitisation.
+type SanitiseOptions struct {
+	NameCharacters       ValidCharacters
+	KeyCharacters        ValidCharacters
+	ValueCharacters      ValidCharacters
+	ReplacementCharacter rune
+}
+
+// Sanitiser sanitises the provided input based on the function executed.
+type Sanitiser interface {
+	// Name sanitises the provided `name` string.
+	Name(n string) string
+
+	// Key sanitises the provided `key` string.
+	Key(k string) string
+
+	// Value sanitises the provided `value` string.
+	Value(v string) string
+}
+
+// NewSanitiser returns a new sanitiser based on provided options.
+func NewSanitiser(opts SanitiseOptions) Sanitiser {
+	return &santiser{
+		nameFn:  opts.NameCharacters.sanitiseFn(opts.ReplacementCharacter),
+		keyFn:   opts.KeyCharacters.sanitiseFn(opts.ReplacementCharacter),
+		valueFn: opts.ValueCharacters.sanitiseFn(opts.ReplacementCharacter),
+	}
+}
+
+// NoOpSanitiseFn returns the input un-touched.
+func NoOpSanitiseFn(v string) string { return v }
+
+// NewNoOpSanitiser returns a sanitiser which returns all inputs un-touched.
+func NewNoOpSanitiser() Sanitiser {
+	return &santiser{
+		nameFn:  NoOpSanitiseFn,
+		keyFn:   NoOpSanitiseFn,
+		valueFn: NoOpSanitiseFn,
+	}
+}
+
+type santiser struct {
+	nameFn  SanitiseFn
+	keyFn   SanitiseFn
+	valueFn SanitiseFn
+}
+
+func (s *santiser) Name(n string) string {
+	return s.nameFn(n)
+}
+
+func (s *santiser) Key(k string) string {
+	return s.keyFn(k)
+}
+
+func (s *santiser) Value(v string) string {
+	return s.valueFn(v)
+}
+
+func (c *ValidCharacters) sanitiseFn(repChar rune) SanitiseFn {
+	return func(value string) string {
+		var buf *bytes.Buffer
+		for idx, ch := range value {
+			// first check if the provided character is valid
+			validCurr := false
+			for i := 0; !validCurr && i < len(c.Ranges); i++ {
+				for j := c.Ranges[i][0]; j <= c.Ranges[i][1]; j++ {
```

Comment:
Any reason we don't check the bounds directly here? Was thinking something along the lines of

```
if ch >= c.Ranges[i][0] && ch <= c.Ranges[i][1] {
    validCurr = true
}
```